ddtrace/internal/flare.py

```diff
@@ -0,0 +1,194 @@
+                    elif response.status == 400:
+                        log.error(
+                            "Upload failed with 400 status code:(%s) %s", response.reason, response.read().decode()
+                        )
+                except Exception as e:
+                    log.error("Failed to send tracer flare")
+                    raise e
+                finally:
+                    client.close()
+                    # Clean up files regardless of success/failure
+                    self._clean_up_files()
+
+    def _generate_config_file(self, pid: int):
+        config_file = f"{TRACER_FLARE_DIRECTORY}/tracer_config_{pid}.json"
+        try:
+            with open(config_file, "w") as f:
+                tracer_configs = {
+                    "configs": config.__dict__,
+                }
+                json.dump(
+                    tracer_configs,
+                    f,
+                    default=lambda obj: obj.__repr__() if hasattr(obj, "__repr__") else obj.__dict__,
+                    indent=4,
+                )
+        except Exception as e:
+            log.warning("Failed to generate %s: %s", config_file, e)
+            if os.path.exists(config_file):
+                os.remove(config_file)
+
+    def _revert_configs(self):
+        ddlogger = get_logger("ddtrace")
+        handler_to_remove = _get_handler(ddlogger, TRACER_FLARE_FILE_HANDLER_NAME)
+        if handler_to_remove:
+            ddlogger.removeHandler(handler_to_remove)
+            log.debug("ddtrace logs will not be routed to the %s file handler anymore", TRACER_FLARE_FILE_HANDLER_NAME)
+        else:
+            log.debug("Could not find %s to remove", TRACER_FLARE_FILE_HANDLER_NAME)
+        ddlogger.setLevel(self.original_log_level)
+
+    def _generate_payload(self, params: Dict[str, str]) -> Tuple[dict, bytes]:
+        tar_stream = io.BytesIO()
+        with tarfile.open(fileobj=tar_stream, mode="w") as tar:
+            for file_name in os.listdir(TRACER_FLARE_DIRECTORY):
+                flare_file_name = f"{TRACER_FLARE_DIRECTORY}/{file_name}"
+                tar.add(flare_file_name)
+        tar_stream.seek(0)
+
+        newline = "\r\n"
+        boundary = secrets.token_hex(16)
+        body = io.StringIO()
+        for key, value in params.items():
+            body.write("--" + boundary + newline)
+            body.write(f'Content-Disposition: form-data; name="{key}"{newline}{newline}')
+            body.write(f"{value}{newline}")
+
+        body.write("--" + boundary + newline)
+        body.write(f'Content-Disposition: form-data; name="flare_file"; filename="flare.tar"{newline}')
+        body.write(f"Content-Type: application/octet-stream{newline}{newline}")
+        body.write(tar_stream.getvalue().decode() + newline)
+        body.write("--" + boundary + "--")
+        headers = {
+            "Content-Type": "multipart/form-data; boundary=" + boundary,
+            "Content-Length": str(len(body.getvalue().encode())),
```

Comment:
TIL: `str(body.getbuffer().nbytes)` should do the same thing.

And that way you don't have to encode the body twice.

the alternative could be:

```python
body_value = body.getvalue().encode()
headers = {
    "Content-Type": "multipart/form-data; boundary=" + boundary,
    "Content-Length": str(len(body_value)),
}

return headers, body_value
```

Reply:
Ah, looks like I can do that with `io.BytesIO` but not `io.StringIO`. `io.StringIO` only has `getvalue()`, which doesn't have `nbytes`. Let me swap it over to `io.BytesIO`